core/src/main/scala/org/bykn/bosatsu/Matchless.scala

```diff
@@ -359,16 +368,36 @@ object Matchless {
           }
 
           variantOf(pack, cname) match {
-            case (None, size) =>
+            case DataRepr.Struct(size) =>
               // this is a struct, so we check each parameter
               asStruct { pos => GetStructElement(arg, pos, size) }
-            case (Some(vidx), size) =>
+            case DataRepr.NewType =>
+              // this is a struct, so we check each parameter
+              asStruct { pos => GetStructElement(arg, pos, 1) }
+            case DataRepr.Enum(vidx, size) =>
               // if we match the variant, then treat it as a struct
               val variant = GetVariant(arg)
               val vmatch = EqualsInt(variant, vidx)
 
               asStruct(GetEnumElement(arg, vidx, _, size))
                 .map(_.map { case (l0, oi, b) => (l0, andInt(Some(vmatch), oi), b) })
+            case DataRepr.ZeroNat =>
+              Monad[F].pure(NonEmptyList((Nil, Some(EqualsNat(arg, DataRepr.ZeroNat)), Nil), Nil))
+            case DataRepr.SuccNat =>
+              params match {
+                case single :: Nil =>
+                  // if we match, we recur on the inner pattern and prev of current
+                  val check = Some(EqualsNat(arg, DataRepr.SuccNat))
+                  val bind: F[(LocalAnon, Expr)] = makeAnon.map { nm => (LocalAnon(nm), PrevNat(arg)) }
+                  for {
+                    pre <- bind
+                    rest <- doesMatch(pre._1, single)
+                  } yield rest.map { case (preLets, cond, res) => (pre :: preLets, andInt(check, cond), res) }
```

Comment:
this is a little weird and makes me wonder if the whole design has an issue:

we have to compute prev unconditionally, we don't have a way to sequence doing a bit of extraction work here and then follow up below. I guess this is what the side-effecting conditions are for, but we're not using it here, so it seems like it will unconditionally run. I think we may want to rethink the composition.